### PR TITLE
Input_file_handling_robustness

### PR DESCRIPTION
**Pull request by tbody-cfs**
This PR makes the input file handling routines more flexible and robust.

1. Lets you not specify any `grid` entry in the input file. You can now run a scalar case.
2. Let's you not specify any `algorithm` entry in the input file. This makes a `Algorithm.empty` which does nothing at all (neater than handling no algorithms as some special case).
3. As a consequence of allowing scalar cases, adds a minor modification to `sanitize_variable` which works around scalar-valued coordinates (see https://github.com/pydata/xarray/issues/1709). This has the pleasing side-effect of getting the correct dimension name in the regression JSON.
4. Adds `cfspopcon` as an entry point, equivalent to `popcon`. Now you can either `poetry run popcon` or `poetry run cfspopcon` — these do the same thing.
5. Lets you pass dictionaries to `input_file_handling` instead of just filepaths, in case you want to programmatically set input values.
6. Runs `ruff` on `tests` to tidy this up

### CODE REVIEW

File: docs/doc_sources/time_independent_inductances_and_fluxes.ipynb
Comment (tbody-cfs) [inline docs/doc_sources/time_independent_inductances_and_fluxes.ipynb:1]: https://fire.pppl.gov/fusion_library.htm was down — I deleted the internal references.

File: cfspopcon/file_io.py [hunk 1]
```diff
@@ -32,9 +32,13 @@ def sanitize_variable(val: xr.DataArray, key: str) -> xr.DataArray:
...
         if val.size == 1:
-            val = val.item().name
+            if not coord:
+                val = val.item().name
+            else:
+                # Xarray doesn't like single-valued coordinated. See https://github.com/pydata/xarray/issues/1709
+                val = xr.DataArray([val.item().name])
         else:
-            val = xr.DataArray([v.name for v in val.values])
+            val = xr.DataArray([v.name for v in val.values], dims=val.dims)
```
Comment (hassec) [inline cfspopcon/file_io.py:41]: 👍 neat addition!

File: cfspopcon/file_io.py [hunk 2]
```diff
@@ -23,7 +23,7 @@
 ]
 
 
-def sanitize_variable(val: xr.DataArray, key: str) -> xr.DataArray:
+def sanitize_variable(val: xr.DataArray, key: str, coord: bool = False) -> xr.DataArray:
     """Strip units and Enum values from a variable so that it can be stored in a NetCDF file."""
```
Comment (hassec) [inline cfspopcon/file_io.py:27]: Mind adding something to the docstrnig about what `coord:bool` controls?

File: cfspopcon/input_file_handling.py [hunk 3]
```diff
@@ -13,7 +13,7 @@
 
 
 def read_case(
-    case: Union[str, Path], kwargs: Optional[dict[str, str]] = None
+    case: Union[str, Path, dict[str, Any]], kwargs: Optional[dict[str, str]] = None
 ) -> tuple[dict[str, Any], Union[CompositeAlgorithm, Algorithm], dict[str, Any], dict[str, Path]]:
     """Read a yaml file corresponding to a given case.
```
Comment (hassec) [inline cfspopcon/input_file_handling.py:18]: How do you feel about leaving `read_case` targeted at reading a yaml file?

If one wants to do things programmatically, one could directly use the `process_*` functions, right? And then one also doesn't have to first create one large dict with all config, to pass it into `read_case`.

I've got an example notebook that does something like that, which I once created to show that workflow to another cfspopcon user. I'd be happy to update it a bit to have it as an example in the docs.

File: docs/doc_sources/time_independent_inductances_and_fluxes.ipynb
Comment (hassec) [reply docs/doc_sources/time_independent_inductances_and_fluxes.ipynb:1]: How about we keep the references, but if they are causing trouble we can add them to the linkcheck ignores, such that they don't fail the docs build.

File: pyproject.toml [hunk 4]
```diff
@@ -19,6 +19,7 @@ classifiers = [
 
 [tool.poetry.scripts]
 popcon = 'cfspopcon.cli:run_popcon_cli'
+cfspopcon = 'cfspopcon.cli:run_popcon_cli'
```
Comment (hassec) [inline pyproject.toml:22]: good idea  👍

**Review by hassec — CHANGES_REQUESTED** [review]
Looks good, just some small comments. 

Regarding the `Alogrithm.empty()`, I'm not sure if I follow how it would be used. If you have a case that doesn't specify any algorithms, what would you do with that case?

File: docs/doc_sources/time_independent_inductances_and_fluxes.ipynb
Comment (tbody-cfs) [reply docs/doc_sources/time_independent_inductances_and_fluxes.ipynb:1]: Done

File: cfspopcon/file_io.py (hunk 2 above)
Comment (tbody-cfs) [reply cfspopcon/file_io.py]: Done

**Comment by tbody-cfs** [comment]
`Algorithm.empty()` is useful if you want to set up a dataset based on an input file. Often useful in a development step, where you want to load up a case and then develop an algorithm in a jupyter notebook.

File: cfspopcon/input_file_handling.py (hunk 3 above)
Comment (tbody-cfs) [reply cfspopcon/input_file_handling.py]: I've split `process_input_dictionary` from `read_case` — so now can choose to use the yaml read and processing, or to just apply the processing.

File: cfspopcon/input_file_handling.py [hunk 5]
```diff
@@ -22,43 +22,64 @@ def read_case(
...
+    return process_input_dictionary(repr_d, case_dir)
+
+
+def process_input_dictionary(
+    repr_d: dict[str, Any], case_dir: Path
+) -> tuple[dict[str, Any], Union[CompositeAlgorithm, Algorithm], dict[str, Any], dict[str, Path]]:
+    """Convert an input dictionary into an processed dictionary, a CompositeAlgorithm and dictionaries defining points and plots.
+
+    Several processing steps are applied, including;
+        * The `algorithms` entry is converted into a `cfspopcon.CompositeAlgorithm` which we'll talk about later. This basically gives the list of operations that we want to perform on the input data.
```
Comment (hassec) [inline cfspopcon/input_file_handling.py:52]:
```suggestion
        * The `algorithms` entry is converted into a `cfspopcon.CompositeAlgorithm`. This basically gives the list of operations that we want to perform on the input data.
```

**Review by hassec — APPROVED** [review]
Just one more tiny comment, otherwise looks great 👍

File: cfspopcon/input_file_handling.py [hunk 6]
```diff
@@ -22,43 +22,67 @@ def read_case(
...
+        * The `points` entry is stored in a separate dictionary. This gives a set of key-value pairs of 'optimal' points (for instance, giving the point with the maximum fusion power gain).
+        * The `grids` entry is converted into an `xr.DataArray` storing a `np.linspace` or `np.logspace` of values which we scan over. We usually scan over `average_electron_density` and `average_electron_temp`, but there's nothing preventing you from scanning over other numerical input variables or having more than 2 dimensions which you scan over (n.b. this can get expensive!).
+        * Each input variable is checked to see if its name matches one of the enumerators in `cfspopcon.named_options`. These are used to store switch values, such as `cfspopcon.named_options.ReactionType.DT` which indicates that we're interested in the DT fusion reaction.
+        * Each input variable is converted into its default units. Default units are retrieved via the `cfspopcon.unit_handling.default_unit` function. This will set, for instance, the `average_electron_temp` values to have units of `keV`.
+    Args:
```
Comment (hassec) [inline cfspopcon/input_file_handling.py:57]:
```suggestion

    Args:
```

Fixing ci

**Review by hassec — APPROVED** [review]

**Pull request merged**